File: packages/core-ts/src/crypto/aes.ts
export interface EncryptResult {
  ciphertext: Uint8Array
  iv: Uint8Array
  tag: Uint8Array
}

export async function encrypt(key: CryptoKey, plaintext: string): Promise<EncryptResult> {
  const iv = crypto.getRandomValues(new Uint8Array(12))
  const encoded = new TextEncoder().encode(plaintext)
  const encrypted = await crypto.subtle.encrypt({ name: 'AES-GCM', iv }, key, encoded)

  // WebCrypto appends 16-byte GCM tag to ciphertext
  const combined = new Uint8Array(encrypted)
  const ciphertext = combined.slice(0, combined.length - 16)
  const tag = combined.slice(combined.length - 16)

  return { ciphertext, iv, tag }
}

export async function decrypt(
  key: CryptoKey,
  ciphertext: Uint8Array,
  iv: Uint8Array,
  tag?: Uint8Array
): Promise<string> {
  let data: Uint8Array<ArrayBuffer>
  if (tag) {
    data = new Uint8Array(ciphertext.length + tag.length)
    data.set(ciphertext, 0)
    data.set(tag, ciphertext.length)
  } else {
    data = new Uint8Array(ciphertext)
  }
  const ivCopy = new Uint8Array(iv)
  const decrypted = await crypto.subtle.decrypt({ name: 'AES-GCM', iv: ivCopy }, key, data)
  return new TextDecoder().decode(decrypted)
}

File: packages/core-py/src/mcp_relay_core/crypto/aes.py
"""AES-256-GCM encryption and decryption."""

import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def encrypt(key: bytes, plaintext: str) -> tuple[bytes, bytes, bytes]:
    """Encrypt plaintext with AES-256-GCM.

    Args:
        key: 32-byte AES key.
        plaintext: String to encrypt.

    Returns:
        Tuple of (ciphertext, iv, tag) where:
        - ciphertext: encrypted data (without tag)
        - iv: 12-byte initialization vector
        - tag: 16-byte authentication tag
    """
    iv = os.urandom(12)
    aesgcm = AESGCM(key)
    # AESGCM.encrypt returns ciphertext + 16-byte tag appended
    combined = aesgcm.encrypt(iv, plaintext.encode("utf-8"), None)
    ciphertext = combined[:-16]
    tag = combined[-16:]
    return ciphertext, iv, tag


def decrypt(key: bytes, ciphertext: bytes, iv: bytes, tag: bytes) -> str:
    """Decrypt ciphertext with AES-256-GCM.

    Args:
        key: 32-byte AES key.
        ciphertext: Encrypted data (without tag).
        iv: 12-byte initialization vector.
        tag: 16-byte authentication tag.

    Returns:
        Decrypted plaintext string.

    Raises:
        cryptography.exceptions.InvalidTag: If authentication fails.
    """
    aesgcm = AESGCM(key)
    # AESGCM.decrypt expects ciphertext + tag concatenated
    combined = ciphertext + tag
    plaintext_bytes = aesgcm.decrypt(iv, combined, None)
    return plaintext_bytes.decode("utf-8")
